### PR TITLE
Add startupProbe result handling to kuberuntime

### LINKED ISSUE #84178: Pod with container with failed StartupProbe stays in Ready: false.
**What happened**:

Pod with container with failed `StartupProbe` stays in `Ready: false`.

**What you expected to happen**:

Pod/container should be killed and restarted and eventually enter `CrashLoopBackOff`.

**How to reproduce it (as minimally and precisely as possible)**:

Start cluster with feature gate `StartupProbe=true`.

Deploy pod:
```yaml
apiVersion: v1
kind: Pod
metadata:
  name: startup-probe-pod
spec:
  containers:
    - image: fedora
      command:
        - sleep
        - inf
      name: example1
      readinessProbe:
        exec:
          command:
            - "true"
        failureThreshold: 2
        periodSeconds: 2
      livenessProbe:
        exec:
          command:
            - "true"
        failureThreshold: 2
        periodSeconds: 2
      startupProbe:
        httpGet:
          path: /healthz
          port: 8080
        failureThreshold: 2
        periodSeconds: 2
```

Watch output:

```
> kubectl get pod
NAME                READY   STATUS    RESTARTS   AGE
startup-probe-pod   0/1     Running   0          3m28s
> kubectl describe pod
Name:         startup-probe-pod                                                                                                                                             
Namespace:    default                                                                                                                                                       
Priority:     0                                                                                                                                                             
Node:         kind-control-plane/172.17.0.2                                                                                                                                 
Start Time:   Tue, 22 Oct 2019 09:54:33 +0200                                                                                                                               
Labels:       <none>                                                                                                                                                        
Annotations:  kubectl.kubernetes.io/last-applied-configuration:                                                                                                             
                {"apiVersion":"v1","kind":"Pod","metadata":{"annotations":{},"name":"startup-probe-pod","namespace":"default"},"spec":{"containers":[{"com...               
Status:       Running                                                                                                                                                       
IP:           10.244.0.2                                                                                                                                                    
IPs:                                                                                                                                                                        
  IP:  10.244.0.2                                                                                                                                                           
Containers:                                                                                                                                                                 
  example1:                                                                                                                                                                 
    Container ID:  containerd://8451c1b089635e2fde65e59a8b00813db728cb7ad8942bf1a81ab0279496b4d9                                                                            
    Image:         fedora                                                                                                                                                   
    Image ID:      docker.io/library/fedora@sha256:8a91dbd4b9d283ca1edc2de5dbeef9267b68bb5dae2335ef64d2db77ddf3aa68                                                         
    Port:          <none>                                                                                                                                                   
    Host Port:     <none>                                                                                                                                                   
    Command:                                                                                                                                                                
      sleep
      inf
    State:          Running
      Started:      Tue, 22 Oct 2019 09:54:54 +0200
    Ready:          False
    Restart Count:  0
    Liveness:       exec [true] delay=0s timeout=1s period=2s #success=1 #failure=2
    Readiness:      exec [true] delay=0s timeout=1s period=2s #success=1 #failure=2
    Startup:        http-get http://:8080/healthz delay=0s timeout=1s period=2s #success=1 #failure=2
    Environment:    <none>
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-s8pdb (ro)
Conditions:
  Type              Status
  Initialized       True 
  Ready             False 
  ContainersReady   False 
  PodScheduled      True 
Volumes:
  default-token-s8pdb:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  default-token-s8pdb
    Optional:    false
QoS Class:       BestEffort
Node-Selectors:  <none>
Tolerations:     node.kubernetes.io/not-ready:NoExecute for 300s
                 node.kubernetes.io/unreachable:NoExecute for 300s
Events:
  Type     Reason            Age                    From                         Message
  ----     ------            ----                   ----                         -------
  Warning  FailedScheduling  <unknown>              default-scheduler            0/1 nodes are available: 1 node(s) had taints that the pod didn't tolerate.
  Normal   Scheduled         <unknown>              default-scheduler            Successfully assigned default/startup-probe-pod to kind-control-plane
  Normal   Pulling           3m54s                  kubelet, kind-control-plane  Pulling image "fedora"
  Normal   Pulled            3m36s                  kubelet, kind-control-plane  Successfully pulled image "fedora"
  Normal   Created           3m33s                  kubelet, kind-control-plane  Created container example1
  Normal   Started           3m33s                  kubelet, kind-control-plane  Started container example1
  Warning  Unhealthy         3m30s (x2 over 3m32s)  kubelet, kind-control-plane  Startup probe failed: Get http://10.244.0.2:8080/healthz: dial tcp 10.244.0.2:8080: connect
: connection refused

```

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"17+", GitVersion:"v1.17.0-alpha.1", GitCommit:"0960c74c3788b1724bd7e7b9933bc49c7e5b5afa", GitTreeState:"clean", BuildDate:"20
19-10-02T23:24:03Z", GoVersion:"go1.12.10", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.2", GitCommit:"c97fe5036ef3df2967d086711e6c0c405941e14b", GitTreeState:"clean", BuildDate:"2019-10-16T
07:13:29Z", GoVersion:"go1.12.10", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration: `kind`
- OS (e.g: `cat /etc/os-release`): `Arch linux
- Kernel (e.g. `uname -a`): `Linux xps13 5.3.7-arch1-1-ARCH #1 SMP PREEMPT Fri Oct 18 00:17:03 UTC 2019 x86_64 GNU/Linux`
- Install tools: `kind`
```yaml
# this config file contains all config fields with comments
kind: Cluster
apiVersion: kind.sigs.k8s.io/v1alpha3
# patch the generated kubeadm config with some extra settings
kubeadmConfigPatches:
- |
  apiVersion: kubeadm.k8s.io/v1beta2
  kind: ClusterConfiguration
  metadata:
    name: config
  apiServer:
    extraArgs:
      "feature-gates": "StartupProbe=true"
  scheduler:
    extraArgs:
      "feature-gates": "StartupProbe=true"
  controllerManager:
    extraArgs:
      "feature-gates": "StartupProbe=true"
- |
  apiVersion: kubeadm.k8s.io/v1beta2
  kind: InitConfiguration
  metadata:
    name: config
  nodeRegistration:
    kubeletExtraArgs:
      "feature-gates": "StartupProbe=true"
# 1 control plane node and 3 workers
nodes:
- role: control-plane
```
- Network plugin and version (if this is a network-related bug): 
- Others:

### PR DESCRIPTION
**Pull request by matthyx**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
StartupProbe failure of a container should (by design) trigger a restart of the Pod, as it's the case for livenessProbes.

**Which issue(s) this PR fixes**:
Fixes #84178

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

### CODE REVIEW

**Comment by odinuge** [comment]
Thanks!

From the way I read the docs, the KEP and the changes in this PR, I think the comments in the tests should be updated. Isn't the correct behavior to restart the container when the startupProbe fails (not depending on the result of the liveness probe)?
https://github.com/kubernetes/kubernetes/blob/ea4570a412a918e24a89992c0daa80efda4a7ae5/test/e2e_node/startup_probe_test.go#L88-L93

---

More detailed comment here: https://github.com/kubernetes/kubernetes/issues/84178#issuecomment-545786090

File: pkg/kubelet/kubelet.go
Comment (odinuge) [inline pkg/kubelet/kubelet.go]: I think this will "race" (not in the sense of a memory race, but the fact that only one of them will get updates to the channel) against this update mechanism, and ignore Success results:
https://github.com/kubernetes/kubernetes/blob/323f99ea8c4faf44090b7554fbfaad2d25120b65/pkg/kubelet/prober/prober_manager.go#L305-L310

Not sure if that is intended or not..

Comment (matthyx) [reply pkg/kubelet/kubelet.go]: But the channel isn't shared... one listens from `startupManager.Updates` and the other from `livenessManager.Updates`. (maybe I don't get the point)

**Comment by matthyx** [comment]
> Isn't the correct behavior to restart the container when the startupProbe fails (not depending on the result of the liveness probe)?

Normally in that test we should have `startupProbe` succeed very quickly, to make sure `livenessProbe` executes and fails - causing the restart.

**Comment by odinuge** [comment]
> Normally in that test we should have startupProbe succeed very quickly, to make sure livenessProbe executes and fails - causing the restart.

The command is `cmd := []string{"/bin/sh", "-c", "sleep 600"}`, and the startupProbe checks for the file `/tmp/health`, and will never succeed. From what I understand from the KEP the container should restart when the Probe fails for the `n`th time, no matter what the result of the livenessProbe is. However the comment `Liveness probe MUST fail after startup probe expires` makes me think that if the livenessProbe succeeds, the container should _not_ restart.

**Comment by matthyx** [comment]
> should a pod like this restart or just start normally

restart and crash loop backoff

File: pkg/kubelet/kubelet.go
Comment (matthyx) [reply pkg/kubelet/kubelet.go]: I hope so... but I have never really worked on that code...

**Comment by matthyx** [comment]
> Isn't the correct behavior to restart the container when the startupProbe fails

Actually the difference between the two tests is the `FailureThreshold` of `startupProbe`... in the first case it is very big (60) to make sure we actually inhibit `livenessProbe` during the short lifespan of the test.
The second case has a bug, because I would expect `startupProbe` to succeed, allowing `livenessProbe` to run and fail, killing the container.
I think I should use `/bin/true` and `/bin/false` to make that more clear and correct.

**Comment by matthyx** [comment]
@odinuge thanks for the comments on e2e tests... here is a PR to fix the ambiguities:
https://github.com/kubernetes/kubernetes/pull/84291

**Review by mattjmcnaughton — COMMENTED** [review]
/uncc

Deferring to @odinuge, who is already doing a great job with the review :)

**Comment by odinuge** [comment]
Looks like there still are some issues with the kubelet killing containers before the `FailureThreashold` is exceeded in the StartupProbe.

To run the test you can run:
```bash
GOOGLE_APPLICATION_CREDENTIALS=google-creds.json make test-e2e-node PARALLELISM=1 REMOTE=true PREEMPTIBLE_INSTANCES=false  TEST_ARGS='--feature-gates=StartupProbe=true' FOCUS="\[Serial\].*\[NodeAlphaFeature:.+\]" SKIP=""
```

More info can be found here: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-node/e2e-node-tests.md

**Comment by matthyx** [comment]
> Looks like there still are some issues with the kubelet killing containers before the `FailureThreashold` is exceeded in the StartupProbe.

Thanks @odinuge ! It's very difficult to test stuff behind my corporate proxy... I will try to find time at home.

**Comment by odinuge** [comment]
> Thanks @odinuge ! It's very difficult to test stuff behind my corporate proxy... I will try to find time at home.

Aha. Should be possible to run locally too `REMOTE=false`, but I have not had any good luck with that.

**Comment by matthyx** [comment]
> Aha. Should be possible to run locally too `REMOTE=false`, but I have not had any good luck with that.

Same here... I have been fighting for 2h now but no luck.

File: pkg/kubelet/prober/results/results_manager.go
Comment (matthyx) [inline pkg/kubelet/prober/results/results_manager.go]: @odinuge is that what you meant? Of course I have no way of  testing that from work :-(

Comment (odinuge) [reply pkg/kubelet/prober/results/results_manager.go]: Yes! So as long as this doesn't have any unintended side effects anywhere else (the constants are exported), this looks good to me, but ill defer that to someone with more knowledge about the code than me.  There are at least some tests that assume `Result` is a bool.

After merging https://github.com/kubernetes/kubernetes/pull/84179 and https://github.com/kubernetes/kubernetes/pull/84291 into this branch, all the e2e tests pass! :tada: 

The tests also pass when reverting the new code block below `case updateStartup := <-kl.startupManager.Updates():` in `pkg/kubelet/kubelet.go`, so I think that can be removed.

The new test suite is also merged and can be found here: https://testgrid.k8s.io/sig-node-kubelet#node-kubelet-serial-alpha. It does however suffer the same issue as the other node-e2e test suits atm., and that is being tracked here: https://github.com/kubernetes/kubernetes/issues/84236

Comment (matthyx) [reply pkg/kubelet/prober/results/results_manager.go]: I am cleaning up the unit tests...
Regarding #84179 and #84291, should I put everything in this branch?

Comment (odinuge) [reply pkg/kubelet/prober/results/results_manager.go]: I think we can keep them as they are, in order to make reviewing a bit simpler.

**Comment by odinuge** [comment]
This, together with #84179 and #84291, will fix #82747.

**Comment by matthyx** [comment]
Thanks @odinuge ! I will try to find enough time tomorrow to fix the tests here and we should be good to go.
Maybe then we could see if the new probe has an impact on scalability...

**Comment by matthyx** [comment]
> The tests also pass when reverting the new code block below case updateStartup := <-kl.startupManager.Updates(): in pkg/kubelet/kubelet.go, so I think that can be removed.

Are you sure?

**Comment by odinuge** [comment]
> Are you sure?

Yes,

This can (and should) be removed:

```go
            case updateStartup := <-kl.startupManager.Updates():
		if updateStartup.Result == proberesults.Failure {
			// The startup manager detected a failure; sync the pod.

			// We should not use the pod from startupManager, because it is never updated after
			// initialization.
			pod, ok := kl.podManager.GetPodByUID(updateStartup.PodUID)
			if !ok {
				// If the pod no longer exists, ignore the update.
				klog.V(4).Infof("SyncLoop (container unhealthy - startup): ignore irrelevant update: %#v", updateStartup)
				break
			}
			klog.V(1).Infof("SyncLoop (container unhealthy - startup): %q", format.Pod(pod))
			handler.HandlePodSyncs([]*v1.Pod{pod})				handler.HandlePodSyncs([]*v1.Pod{pod})
```

**Comment by matthyx** [comment]
Squashed and ready for review! Thanks a lot for all your help @odinuge !
/assign @Random-Liu

File: pkg/kubelet/prober/results/results_manager.go
Comment (odinuge) [reply pkg/kubelet/prober/results/results_manager.go]: Ill defer the naming of this to the approvers (eg. `Pending`)

File: pkg/kubelet/prober/results/results_manager.go
Comment (odinuge) [inline pkg/kubelet/prober/results/results_manager.go]: Should we add the new `Result` type here (and in `ToPrometheusType`) as well? I think the current behavior is ok, reporting -1/"UNKNOWN", but it may be nice to be explicit. Ill defer that to the approvers :smile:

**Review by odinuge — APPROVED** [review]
Thanks for the work @matthyx! Hopefully we can get this into beta in v1.17!

/lgtm

**Comment by matthyx** [comment]
@Random-Liu, @dchen1107, @derekwaynecarr, @tallclair, @vishh, @yujuhong this PR is really needed for 1.17, can you please make sure we don't miss code-freeze?

File: pkg/kubelet/kuberuntime/kuberuntime_manager_test.go
Comment (Random-Liu) [inline pkg/kubelet/kuberuntime/kuberuntime_manager_test.go]: Can we `createTestRuntimeManager` inside the loop? So that we don't need to reset the status.

File: pkg/kubelet/prober/results/results_manager.go
Comment (dchen1107) [reply pkg/kubelet/prober/results/results_manager.go]: One small concern I have is I am not sure if anyone already depend on the previous defined results. cc/ @yujuhong

File: pkg/kubelet/prober/results/results_manager.go
Comment (Random-Liu) [inline pkg/kubelet/prober/results/results_manager.go]: The startup probe logic is almost the same with liveness probe, right?

Why Success/Failure is sufficient for liveness probe, but not startup probe?

File: pkg/kubelet/prober/worker_test.go
Comment (yujuhong) [inline pkg/kubelet/prober/worker_test.go]: remove "v1"

File: pkg/kubelet/prober/results/results_manager.go
Comment (yujuhong) [inline pkg/kubelet/prober/results/results_manager.go]: nit: just use iota?
```
const (
  Unknown Result = iota -1
  Success
  Failure
)
```

File: pkg/kubelet/prober/results/results_manager.go
Comment (Random-Liu) [reply pkg/kubelet/prober/results/results_manager.go]: OK. Because the init values of liveness probe and startup probes are different.

Comment (Random-Liu) [reply pkg/kubelet/prober/results/results_manager.go]: Given that `Unknown` is newly introduced, and only used for initializing startup probe, I think this should be safe.

File: pkg/kubelet/kuberuntime/kuberuntime_manager_test.go
Comment (matthyx) [reply pkg/kubelet/kuberuntime/kuberuntime_manager_test.go]: Makes sense... maybe it could go into another PR?

**Review by odinuge — APPROVED** [review]
/lgtm

**Comment by matthyx** [comment]
@dchen1107 I think you can **approve** when you're back!

**Comment by matthyx** [comment]
Thanks @dchen1107

**Pull request merged**
